src/compas_fab/backends/pybullet/client.py

```diff
@@ -0,0 +1,410 @@
+from __future__ import absolute_import
+from __future__ import division
+from __future__ import print_function
+
+import os
+import tempfile
+from itertools import combinations
+
+import compas
+from compas._os import system
+from compas.robots import RobotModel
+
+from compas_fab.backends.interfaces.client import ClientInterface
+from compas_fab.robots import Robot
+from compas_fab.utilities import LazyLoader
+
+from .const import BASE_LINK_ID
+from .const import BodyInfo
+from .const import GREY
+from .const import JointInfo
+from .const import LinkState
+from .const import NULL_ID
+from .const import RED
+from .const import STATIC_MASS
+from .const import ZERO_FRAME
+from .planner import PyBulletPlanner
+from .utils import LOG
+from .utils import redirect_stdout
+from .conversions import frame_from_pose
+from .conversions import pose_from_frame
+
+pybullet = LazyLoader('pybullet', globals(), 'pybullet')
+
+
+__all__ = [
+    'PyBulletClient',
+]
+
+
+class PyBulletBase(object):
+    def __init__(self, use_gui):
+        self.client_id = None
+        self.use_gui = use_gui
+
+    def connect(self, shadows=True, color=None, width=None, height=None):
+        # Shared Memory: execute the physics simulation and rendering in a separate process
+        # https://github.com/bulletphysics/bullet3/blob/master/examples/pybullet/examples/vrminitaur.py#L7
+        self.detect_display()
+        method = pybullet.GUI if self.use_gui else pybullet.DIRECT
+        options = self.compose_options(color, width, height)
+        with redirect_stdout():
+            self.client_id = pybullet.connect(method, options=options)
+        if 0 > self.client_id:
+            raise Exception('Error in establishing connection with PyBullet.')
+        if self.use_gui:
+            self.configure_debug_visualizer(shadows)
+
+    def detect_display(self):
+        if self.use_gui and system != 'darwin' and not compas.is_windows() and ('DISPLAY' not in os.environ):
```

Comment:
I had assumed that this was to prevent troubles when running PyBullet on a headless server without openGL support, that is to say, not Travis related. But perhaps we should expect the user to deal with that and delete this check. This is adapted from @yijiangh 's work, so maybe he can explain more.